lib/pos_consensus.go

```diff
@@ -0,0 +1,685 @@
+		return errors.Errorf("FastHotStuffConsensus.HandleLocalTimeoutEvent: Error signing validator timeout: %v", err)
+	}
+
+	// Process the timeout message locally in the FastHotStuffEventLoop
+	if err := cc.fastHotStuffEventLoop.ProcessValidatorTimeout(timeoutMsg); err != nil {
+		// This should never happen. If we error here, it means that the timeout message is stale
+		// beyond the committed tip, the timeout message is malformed, or the timeout message is
+		// is duplicated for the same view. In any case, something is very wrong. We should not
+		// broadcast this message to the network.
+		return errors.Errorf("FastHotStuffConsensus.HandleLocalTimeoutEvent: Error processing timeout locally: %v", err)
+	}
+
+	// Broadcast the timeout message to the network
+	// TODO: Broadcast the timeout message to the network or alternatively to just the block proposer
+
+	return nil
+}
+
+// HandleValidatorTimeout is called when we receive a validator timeout message from a peer. This function
+// processes the timeout locally in the FastHotStuffEventLoop.
+func (cc *FastHotStuffConsensus) HandleValidatorTimeout(pp *Peer, msg *MsgDeSoValidatorTimeout) error {
+	// No need to hold a lock on the consensus because this function is a pass-through
+	// for the FastHotStuffEventLoop which guarantees thread-safety for its callers.
+
+	// Process the timeout message locally in the FastHotStuffEventLoop
+	if err := cc.fastHotStuffEventLoop.ProcessValidatorTimeout(msg); err != nil {
+		// If we can't process the timeout locally, then it must somehow be malformed, stale,
+		// or a duplicate vote/timeout for the same view.
+		return errors.Wrapf(err, "FastHotStuffConsensus.HandleValidatorTimeout: Error processing timeout: ")
+	}
+
+	// Happy path
+	return nil
+}
+
+func (cc *FastHotStuffConsensus) HandleBlock(pp *Peer, msg *MsgDeSoBlock) error {
+	// Hold a lock on the consensus, because we will need to mutate the Blockchain
+	// and the FastHotStuffEventLoop data structures.
+	cc.lock.Lock()
+	defer cc.lock.Unlock()
+
+	if !cc.fastHotStuffEventLoop.IsRunning() {
+		return errors.Errorf("FastHotStuffConsensus.HandleBlock: FastHotStuffEventLoop is not running")
+	}
+
+	// Try to apply the block as the new tip of the blockchain. If the block is an orphan, then
+	// we will get back a list of missing ancestor block hashes. We can fetch the missing blocks
+	// from the network and retry.
+	missingBlockHashes, err := cc.tryProcessBlockAsNewTip(msg)
+	if err != nil {
+		// If we get an error here, it means something went wrong with the block processing algorithm.
+		// Nothing we can do to recover here.
+		return errors.Errorf("FastHotStuffConsensus.HandleBlock: Error processing block as new tip: %v", err)
+	}
+
+	// If there are missing block hashes, then we need to fetch the missing blocks from the network
+	// and retry processing the block as a new tip. We'll request the blocks from the same peer.
+	if len(missingBlockHashes) > 0 {
+		pp.QueueMessage(&MsgDeSoGetBlocks{
```

Comment:
This seems like it should totally work, but a thought I had is it may be inefficient in the sense that you will do a lot of round-trips if you get a far-ahead block. For example, if the peer has `A -> B -> C -> D -> E -> F` and you're at just `A` and the peer sends you `F`, then you will try to connect F, fail, then request E, then you'll get E, try to connect it fail and request D, and so on. I don't think we need to optimize this right now because nodes will generally be nearly-synced at this stage. But leave a comment that this behavior exists if it's correct.

One way to fix this is to send your tip inside the GetBlocks message, and then the node can auto-send you everything from your tip to its tip. But again I think this is probably a premature optimization we don't need to make right now. If you like that optimization then leave it as a TODO for fun.

Reply:
Btw just added context: The above suggestion is a simplified version of what GetHeaders does. The GetHeaders thing does a sophisticated "common ancestor" check between the two nodes by sending an exponentially-backed-off list of your headers, eg: `tip, tip-1, tip-2, tip-4, tip-8, tip-16, ..., genesis`. This covers cases where nodes are super divergent, which can happen in Bitcoin but can't happen in PoS so it doesn't make sense to port it over. Instead, sending the tip only, and then having the node send you its next N blocks from the tip automatically, seems like the "most right" thing to do with PoS. But again I think what you have works and just adding a TODO is good and will work great.

Reply:
Added the comment here: https://github.com/deso-protocol/core/pull/963